core/dragging/block_drag_strategy.ts

```diff
@@ -0,0 +1,395 @@
+import {config} from '../config.js';
+import type {BlockMove} from '../events/events_block_move.js';
+import {finishQueuedRenders} from '../render_management.js';
+import * as layers from '../layers.js';
+
+/** Represents a nearby valid connection. */
+interface ConnectionCandidate {
+  /** A connection on the dragging stack that is compatible with neighbour. */
+  local: RenderedConnection;
+
+  /** A nearby connection that is compatible with local. */
+  neighbour: RenderedConnection;
+
+  /** The distance between the local connection and the neighbour connection. */
+  distance: number;
+}
+
+export class BlockDragStrategy implements IDragStrategy {
+  private workspace: WorkspaceSvg;
+
+  /** The parent block at the start of the drag. */
+  private startParentConn: RenderedConnection | null = null;
+
+  /**
+   * The child block at the start of the drag. Only gets set if
+   * `healStack` is true.
+   */
+  private startChildConn: RenderedConnection | null = null;
+
+  private startLoc: Coordinate | null = null;
+
+  private connectionCandidate: ConnectionCandidate | null = null;
+
+  private connectionPreviewer: IConnectionPreviewer | null = null;
+
+  private dragging = false;
+
+  constructor(private block: BlockSvg) {
+    this.workspace = block.workspace;
+  }
+
+  /** Returns true if the block is currently movable. False otherwise. */
+  isMovable(): boolean {
+    return (
+      this.block.isOwnMovable() &&
+      !this.block.isShadow() &&
+      !this.block.isDeadOrDying() &&
+      !this.workspace.options.readOnly
+    );
+  }
+
+  /**
+   * Handles any setup for starting the drag, including disconnecting the block
+   * from any parent blocks.
+   */
+  startDrag(e?: PointerEvent): void {
+    this.dragging = true;
+    if (!eventUtils.getGroup()) {
+      eventUtils.setGroup(true);
+    }
+    this.fireDragStartEvent_();
+
+    this.startLoc = this.block.getRelativeToSurfaceXY();
+
+    const previewerConstructor = registry.getClassFromOptions(
+      registry.Type.CONNECTION_PREVIEWER,
+      this.workspace.options,
+    );
+    this.connectionPreviewer = new previewerConstructor!(this.block);
+
+    // During a drag there may be a lot of rerenders, but not field changes.
+    // Turn the cache on so we don't do spurious remeasures during the drag.
+    dom.startTextWidthCache();
+    this.workspace.setResizesEnabled(false);
+    blockAnimation.disconnectUiStop();
+
+    const healStack = !!e && (e.altKey || e.ctrlKey || e.metaKey);
+
+    if (this.shouldDisconnect_(healStack)) {
+      this.disconnectBlock_(healStack);
+    }
+    this.block.setDragging(true);
+    this.workspace.getLayerManager()?.moveToDragLayer(this.block);
+  }
+
+  /**
+   * Whether or not we should disconnect the block when a drag is started.
+   *
+   * @param healStack Whether or not to heal the stack after disconnecting.
+   * @returns True to disconnect the block, false otherwise.
+   */
+  private shouldDisconnect_(healStack: boolean): boolean {
```

Comment:
Nit: since these functions are private, do they still need underscores at the end of the names?

Reply:
Ope good point!